### ISSUE TITLE
Reversed order of ReLU and BatchNorm during backward propagation

### ISSUE THREAD

**Issue by jih332** [pytorch/vision#1123]
### Environment:
GoogleColab default env
torch: 1.1.0
torchvision: 0.3.0
Python: 3.6.8 (default, Jan 14 2019, 11:02:34) 
[GCC 8.0.1 20180414 (experimental) [trunk revision 259383]]
GPU: True

### Description:
Order of ReLU and its preceding layer is reversed during backward propagation.
#### Code:
the function `is_supported_instance()` below checks whether a module is one of linear/conv/pooling/batchnorm/relu layers.
```python3
def fw_hook(module, input, output):
    print(f'forward hook in module {module.__full_name__} {module.__class__.__name__}')
    
    def bw_hook(grad):
        print(f'backward hook in module {module.__full_name__} {module.__class__.__name__}')
        
    handle = output.register_hook(bw_hook)
    module.__bw_hook_handle__ = handle

def register_hook(module, hook_fn):
    if is_supported_instance(module):
        handle = module.register_forward_hook(hook_fn)
        module.__fw_hook_handle__ = handle
    
tm = torchvision.models.resnet18()

for n, m in tm.named_modules():
    m.__full_name__ = n.replace('.', '/')
    
tm.apply(lambda m: register_hook(m, fw_hook))

input = torch.ones([2, 3, 224, 224], dtype = torch.float)
out = tm(input)
print(out.shape)
out = out.mean()
out.backward()
```
#### output:
```
...
forward hook in layer4/1/conv1 type: Conv2d
forward hook in layer4/1/bn1 type: BatchNorm2d
forward hook in layer4/1/relu type: ReLU
forward hook in layer4/1/conv2 type: Conv2d
forward hook in layer4/1/bn2 type: BatchNorm2d
forward hook in layer4/1/relu type: ReLU
forward hook in avgpool type: AdaptiveAvgPool2d
forward hook in fc type: Linear
torch.Size([2, 1000])
backward hook in fc type: Linear
backward hook in avgpool type: AdaptiveAvgPool2d
backward hook in layer4/1/bn2 type: BatchNorm2d
backward hook in layer4/1/relu type: ReLU
backward hook in layer4/1/conv2 type: Conv2d
backward hook in layer4/1/bn1 type: BatchNorm2d
backward hook in layer4/1/relu type: ReLU
backward hook in layer4/1/conv1 type: Conv2d
...
```

so if we follow the order it forward props, the backward prop should be 
```
backward hook in fc type: Linear
backward hook in avgpool type: AdaptiveAvgPool2d
backward hook in layer4/1/relu type: ReLU
backward hook in layer4/1/bn2 type: BatchNorm2d
backward hook in layer4/1/conv2 type: Conv2d
```
however, as indicated in actual output, the relative order of ReLU and BatchNorm was reversed. I've checked `resnet18`, `resnet34`, `resnet50` and `vgg11`, this has always been the case. So can anyone explain a bit why did this happen? Thanks.

**Comment by fmassa** [comment]
This might be due to the fact that the `ReLU` is in-place, and the operation just before the `ReLU` is also in-place (`+=`), and autograd does some extra checks to see if it is safe to compute the gradients or not. Also note that the order of operations during backward is not necessarily the same as during forward, and the autograd engine can reorder the operations.

As such, I'm closing this issue. If you still have questions, I think it might be better to open an issue in PyTorch, because this is something happening inside autograd.
